Streams: transfer infra, and transfer of ReadableStream

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Add transfer support to ReadableStream. Part of https://github.com/servo/servo/issues/34676

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
Implement the transfer of readablestream

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement algos for underlyingsource container

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement pack and post message

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement cross realm transform send error

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
implement pack and post message handling error

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
rename deserialized stored to readable streams

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
move message handlers

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add tranform handlers, restucture port task routing to better match spec, with docs

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
enable ports are part of setting up cross realm transforms
when a port is started, immediately route buffered messages(removing the previous task queueing).
fix typos

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add note about the pair of transforms on the global

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
root transforms

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
set high watermark for readable side of transform to 0, as per the spec

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix typo and resolve cancel algo

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#51932) with upstreamable changes.

gterzian committed:
fix error propagation

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
turn on part of test suite

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
reduce visibility of port methods, remove unnecessary comments
add comment on using port infra for transfer
Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update test expectations and config

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
update structured clone test expectations

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add issue number for todo of disentanglement

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
nit for cross realm send error

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
improve doc on underlying sink type

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
nit on writable controller start doc

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
fix warnings

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
@Taym95 ready for review please.

gterzian committed:
add note about script entry in route task

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on tests/wpt/meta/streams/transferable/reason.html.ini:
```diff
@@ -0,0 +1,3 @@
+[reason.html]
+  [DOMException errors should be preserved]
+    expected: FAIL
```
This is https://github.com/servo/servo/issues/36463

gterzian commented on components/script/dom/writablestreamdefaultcontroller.rs:
```diff
@@ -480,9 +538,23 @@ impl WritableStreamDefaultController {
-                // TODO: implement transfer.
-                Promise::new_resolved(global, cx, (), can_gc)
+            UnderlyingSinkType::Transfer { ref port, .. } => {
+                // Let result be PackAndPostMessageHandlingError(port, "error", reason).
```
yep.

gterzian commented on components/script/dom/writablestream.rs:
```diff
@@ -834,6 +839,58 @@ impl WritableStream {
     }
+
+    /// <https://streams.spec.whatwg.org/#abstract-opdef-setupcrossrealmtransformwritable>
+    pub(crate) fn setup_cross_realm_transform_writable(
```
It supports some of the internal stuff because it's necessary, and adding the rest in another PR will be easy.

gterzian committed:
use cx in get type and value from message,
add link to underlying sink transfer algos,
un-ignore test suites which we can use soon.

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented on components/script/dom/readablestream.rs:
```diff
@@ -1941,6 +1995,149 @@ impl ReadableStreamMethods<crate::DomTypeHolder> for ReadableStream {
+    // Assert: type is a String.
+    let result = unsafe {
+        DOMString::from_jsval(
+            *GlobalScope::get_cx(),
```
yep

gterzian committed:
fix use of dataclone error after rebase

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian commented:
@Taym95 Thanks. I have addressed your review comments, and also rebased and fixed a conflict. Did you meant to do another review? Otherwise this needs an explicit approval and not just a comment.

gterzian committed:
fix clippy

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
un-ignore test using writable stream

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

Taym95 reviewed (approved):
I checked all the steps LGTM, maybe @jdm wants to have a look, else we can merge it!

jdm commented:
Nope, go ahead!

gterzian commented:
@Taym95 @jdm thank you
